lib/Serialization/ModuleFile.cpp

```diff
@@ -1222,13 +1223,164 @@ bool ModuleFile::readModuleDocIfPresent() {
   return true;
 }
 
+class ModuleFile::DeclUSRTableInfo {
+  public:
+  using internal_key_type = StringRef;
+  using external_key_type = StringRef;
+  using data_type = uint32_t;
+  using hash_value_type = uint32_t;
+  using offset_type = unsigned;
+
+  internal_key_type GetInternalKey(external_key_type key) { return key; }
+
+  hash_value_type ComputeHash(internal_key_type key) {
+    assert(!key.empty());
+    return llvm::djbHash(key, SWIFTSOURCEINFO_HASH_SEED);
+  }
+
+  static bool EqualKey(internal_key_type lhs, internal_key_type rhs) { return lhs == rhs; }
+
+  static std::pair<unsigned, unsigned> ReadKeyDataLength(const uint8_t *&data) {
+    unsigned keyLength = endian::readNext<uint32_t, little, unaligned>(data);
+    unsigned dataLength = 4;
+    return { keyLength, dataLength };
+  }
+
+  static internal_key_type ReadKey(const uint8_t *data, unsigned length) {
+    return StringRef(reinterpret_cast<const char*>(data), length);
+  }
+
+  data_type ReadData(internal_key_type key, const uint8_t *data, unsigned length) {
+    assert(length == 4);
+    return *reinterpret_cast<const uint32_t*>(data);
```

Comment:
\*sigh\* Technically this kind of reinterpret_cast is illegal in C++, please use the `endian::readNext` API anyway. (Also, it's important if we ever cross platform boundaries with endianness! For some reason!)